### ISSUE TITLE
A Research Code Base for Deep Reinforcement Learning in PyTorch  The Berkeley Artificial Intelligence Research Blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5538]
A Research Code Base for Deep Reinforcement Learning in PyTorch &ndash; The Berkeley Artificial Intelligence Research Blog<br>
<br>
https://ift.tt/2mn6ZrG<br>
<br>
Daniel Seita<br>
<br>
<div>
<article>
<p>Since the advent of deep reinforcement learning for <a href="https://arxiv.org/abs/1312.5602">game play</a> in 2013, and <a href="https://arxiv.org/abs/1502.05477">simulated robotic control</a> shortly after, a multitude of new algorithms have flourished. Most of these are model-free algorithms which can be categorized into three families: deep Q-learning, policy gradients, and Q-value policy gradients. Because they rely on different learning paradigms, and because they address different (but overlapping) control problems, distinguished by discrete versus continuous action sets, these three families have developed along separate lines of research. Currently, very few if any code bases incorporate all three kinds of algorithms, and many of the original implementations remain unreleased. As a result, practitioners often must develop from different starting points and potentially learn a new code base for each algorithm of interest or baseline comparison. RL researchers must invest time reimplementing algorithms&ndash;a valuable individual exercise but one which incurs redundant effort across the community, or worse, one that presents a barrier to entry.</p>
<p>Yet these algorithms share a great depth of common reinforcement learning machinery. We are pleased to share <em>rlpyt</em>, which leverages this commonality to offer all three algorithm families built on a shared, optimized infrastructure, in one repository. Available from BAIR at <a href="https://github.com/astooke/rlpyt">https://github.com/astooke/rlpyt</a>, it contains modular implementations of many common deep RL algorithms in Python using Pytorch, a leading deep learning library. Among numerous existing implementations, rlpyt is a more comprehensive open-source resource for researchers.</p>
<p>rlpyt is designed as a high-throughput code base for small- to medium-scale research in deep RL (large-scale being <a href="https://deepmind.com/blog/article/alphastar-mastering-real-time-strategy-game-starcraft-ii">DeepMind AlphaStar</a> or <a href="https://openai.com/blog/openai-five/">OpenAI Five</a>, with 100&rsquo;s of GPUs). This blog post briefly introduces its features and relation to prior work. Notably, rlpyt reproduces the recent record-setting results in the Atari domain from &ldquo;Recurrent Experience Replay in Distributed Reinforcement Learning&rdquo; (<a href="https://openreview.net/forum?id=r1lyTjAqYX">R2D2</a>)&ndash;except without requiring distributed compute infrastructure to gather the 10&rsquo;s of billions of frames of gameplay needed. We also introduce a new data structure, called the <em>namedarraytuple</em>, which is used extensively in rlpyt for handling collections of numpy arrays. Further technical discussion, including more detailed implementation and usage notes, are provided in a <a href="https://arxiv.org/abs/1909.01500">white paper posted on arxiv</a>.</p>
<p>Key features and capabilities include the following:</p>
<ul>
<li>Run experiments in serial mode (helpful for debugging, maybe sufficient for experiments).</li>
<li>Run experiments parallelized, with options for parallel sampling and/or multi-GPU optimization.</li>
<li>Sampling and optimization synchronous or asynchronous (via replay buffer).</li>
<li>Use CPU or GPU for training and/or batched action selection during environment sampling.</li>
<li>Full support for recurrent agents.</li>
<li>Online or offline evaluation and logging of agent diagnostics during training.</li>
<li>Launching utilities for stacking / queueing sets of experiments on local computer.</li>
<li>Modularity for easy modification and re-use of existing components.</li>
<li>Compatible with OpenAI Gym environment interface.</li>
</ul>
<p>Algorithms implemented (check the repository for additions):</p>
<ul>
<li>Policy Gradient: <a href="https://arxiv.org/abs/1602.01783">A2C</a>, <a href="https://arxiv.org/abs/1707.06347">PPO</a>
</li>
<li>
<a href="https://arxiv.org/abs/1312.5602">DQN</a> + variants: <a href="https://arxiv.org/abs/1509.06461">Double</a>, <a href="https://arxiv.org/abs/1511.06581">Dueling</a>, <a href="https://arxiv.org/abs/1707.06887">Categorical</a>, <a href="https://arxiv.org/abs/1710.02298">Rainbow</a> minus Noisy Nets, Recurrent (<a href="https://openreview.net/forum?id=r1lyTjAqYX">R2D2-style</a>) (coming soon: <a href="https://arxiv.org/abs/1806.06923">Implicit Quantile Networks</a>)</li>
<li>Q-value Policy Gradients: <a href="https://arxiv.org/abs/1509.02971">DDPG</a>, <a href="https://arxiv.org/abs/1802.09477">TD3</a>, <a href="https://arxiv.org/abs/1801.01290">SAC</a> (coming soon: <a href="https://arxiv.org/abs/1804.08617">Distributional DDPG</a>)</li>
</ul>
<p>Replay buffers (supporting DQN and QPG) are included with the following optional features: n-step returns, <a href="https://arxiv.org/abs/1511.05952">prioritized replay</a>, sequence replay (for recurrence), frame-based buffers (e.g. to store only unique Atari frames from multi-frame observations).</p>
<p><strong>Sampling</strong>. The two phases of model-free RL, sampling environment interactions and training the agent, can be parallelized differently. For sampling, rlpyt includes three basic options: serial, parallel-CPU, and parallel-GPU. Serial sampling is the simplest, as the entire program runs in one Python process, and this is often useful for debugging. But environments are typically CPU-based and single-threaded, so the parallel samplers use worker processes to run environment instances, speeding up the overall collection rate. CPU sampling also runs the agent&rsquo;s neural network in the workers for action selection. GPU sampling batches all environments&rsquo; observations together for action-selection in the master process, for more efficient use of the GPU. These configurations are depicted in the Figure [sampler]. One additional offering is alternating-GPU sampling, which uses two groups of workers: one executes environment simulation while the other awaits new actions. This may provide speedups when the action-selection time is similar to but shorter than the batch environment simulation time.</p>
<p><img src="https://bair.berkeley.edu/static/blog/rlpyt/samplers.png"><br>
<i>Figure [sampler]. Environment interaction sampling schemes. a) Serial: agent and environments execute within one Python process. b) Parallel-CPU: agent and environments execute on CPU in parallel worker processes. c) Parallel-GPU: environments execute on CPU in parallel workers processes, agent executes in central process, enabling batched action-selection.</i></p>
<p><strong>Optimization</strong>. Synchronous multi-GPU optimization is included via PyTorch&rsquo;s <em>DistributedDataParallel</em> wrapper. The entire sampler-optimizer stack is replicated in a separate process for each GPU, and the model implicitly synchronizes by all-reducing the gradient during backpropagation. The DistributedDataParallel tool automatically reduces the gradient in chunks concurrently with backpropagation for better scaling on large networks. This arrangement is shown in Figure [sync]. The sampler can be any serial or parallel configuration described earlier.</p>
<p><img width="600" src="https://bair.berkeley.edu/static/blog/rlpyt/sync.png"><br>
<i>Figure [sync]. Synchronous multi-process reinforcement learning. Each python process runs a copy of the fully sample-algorithm stack, with synchronization enforced implicitly during backpropagation in PyTorch&rsquo;s `DistribuedDataParallel` class. Both GPU (NCCL backend) and CPU (gloo backend) modes are supported.</i></p>
<p><strong>Asynchronous Sampling-Optimization</strong>. In the configurations depicted so far, the sampler and optimizer operate sequentially in the same Python process. In some cases, however, running optimization and sampling asynchronously achieves better hardware utilization, by allowing both to run continuously. This was the case in our R2D2 reproduction and could also be true for learning on real robots. In asynchronous mode, separate Python processes run the training and sampling, tied together by a replay buffer built on shared memory. Sampling runs uninterrupted by the use of a double buffer for data batches, which yet another Python process copies into the main buffer, under write lock. See Figure [async]. The optimizer and sampler may be parallelized independently, perhaps each using a different number of GPUs, to achieve best overall utilization and speed.</p>
<p><img width="500" src="https://bair.berkeley.edu/static/blog/rlpyt/async.png"><br>
<i>Figure [async]. Asynchronous sampling/optimization mode. Separate python processes run optimization and sampling via a shared-memory replay buffer under read-write lock. Memory copier processes write to the replay buffer, freeing the sampler to proceed immediately from batch to batch of collection.</i></p>
<p><strong>Which configuration is best</strong>? For creating or modifying agents, models, algorithms, and environments, serial mode will be the easiest for debugging. Once the serial program runs smoothly, one can easily explore the more sophisticated infrastructures, such as parallel sampling, multi-GPU optimization, and asynchronous sampling, since they are built on largely the same interfaces. The optimal configuration may depend on the learning problem, available computer hardware, and number of experiments to run. Parallelism included in rlpyt is limited to the single-node case, although its components could serve as building blocks for a distributed framework.</p>
<p>We highlight learning curves reproducing the R2D2 result in Atari, which was previously only feasible using distributed computing. This benchmark includes a recurrent agent trained from a replay buffer for on the order of 10 billion samples (40 billion frames). R2D1 (non-distributed R2D2) exercises several of rlpyt&rsquo;s more advanced infrastructure components to achieve this, namely multi-GPU asynchronous sampling mode with the alternating-GPU sampler. In Figure [R2D1], we reproduce several learning curves which surpass any previous algorithm. We do note that results did not reproduce perfectly on all games&ndash;for example in Gravitar the score plateaued at a low level. The white paper contains more discussion on this point, as well as further verifications for the other algorithms and environments.</p>
<p><img src="https://bair.berkeley.edu/static/blog/rlpyt/r2d1_curves_2.png"><br>
<i>Figure [R2D1 curves]. Reproduction of R2D2 learning curves in rlpyt, using a single computer.</i></p>
<p>The original, distributed implementation of R2D2 quoted about 66,000 steps per second (SPS) using 256 CPUs for sampling and 1 GPU for training. rlpyt achieves over 16,000 SPS when using only 24 CPUs (2x Intel Xeon Gold 6126, circa 2017) and 3 Titan-Xp GPUs in a single workstation (one GPU for training, two for action-serving in the alternating sampler). This may be enough to enable experimentation without access to distributed infrastructure. One possibility for future research is to increase the replay ratio (here set to 1) for faster learning using multi-GPU optimization. Figure [Amidar] shows the same learning curve over three different measures: environment steps (i.e. 1 step = 4 frames), model updates, and time. This run reached 8 billion steps and 1 million updates in less than 138 hours.</p>
<p><img width="600" src="https://bair.berkeley.edu/static/blog/rlpyt/amidar_time_2.png"><br>
<i>Figure [Amidar]. The same learning curve over three horizontal axes: environment steps, model updates, and wall-clock time for rlpyt&rsquo;s R2D1 implementation run in asynchronous sampling mode using 24 CPU cores and 3 GPUs.</i></p>
<p>rlpyt introduces new object classes, called namedarraytuples, for easier organization of collections of numpy arrays or torch tensors. A namedarraytuple is essentially a namedtuple which exposes indexed or sliced read/writes into the structure. Consider writing into a (possibly nested) dictionary of arrays which share some common dimensions for addressing:</p>
<div>
<div>
<pre><code><span>for</span> <span>k</span><span>,</span> <span>v</span> <span>in</span> <span>src</span><span>.</span><span>items</span><span>():</span>
  <span>if</span> <span>isinstance</span><span>(</span><span>dest</span><span>[</span><span>k</span><span>],</span> <span>dict</span><span>):</span>
    <span>..</span><span>recurse</span><span>..</span>
  <span>dest</span><span>[</span><span>k</span><span>][</span><span>slice_or_indexes</span><span>]</span> <span>=</span> <span>v</span>
</code></pre>
</div>
</div>
<p>This code is replaced by the following:</p>
<div>
<div>
<pre><code><span>dest</span><span>[</span><span>slice_or_indexes</span><span>]</span> <span>=</span> <span>src</span>
</code></pre>
</div>
</div>
<p>Importantly, the syntax is the same whether <code>dest</code> and <code>src</code> are individual numpy arrays or arbitrarily-structured collections of arrays (the structures of <code>dest</code> and <code>src</code> must match, or <code>src</code> can be a single value to apply to all fields). rlpyt uses this data structure extensively&ndash;different elements of training data are organized with the same leading dimensions, making it easy to interact with desired time- or batch-dimensions. Namedarraytuples also naturally support environments with multi-modal actions or observations (e.g. vision and joint angles). This is useful when different modes interface with the neural network at different layers, as it allows the intermediate infrastructure code to remain unchanged regardless of this structure. For more details, see the code and documentation for namedarraytuples in <a href="https://github.com/astooke/rlpyt/blob/master/rlpyt/utils/collections.py">rlpyt/utils/collections.py</a>.</p>
<p>For newcomers to deep RL, other resources may be better for familiarization with algorithms, such as OpenAI Spinning Up (<a href="https://github.com/openai/spinningup">code</a> / <a href="https://spinningup.openai.com/en/latest/">docs</a>). rlpyt is a revision of <a href="https://github.com/astooke/accel_rl">accel_rl</a>, which explored scaling RL in the Atari domain using <a href="https://arxiv.org/pdf/1605.02688.pdf">Theano</a>, see <a href="https://arxiv.org/abs/1803.02811">the paper</a> for results. For a further study of batch-size scaling in deep learning and RL, see <a href="https://arxiv.org/abs/1812.06162">this OpenAI report</a>. rlpyt and accel_rl were originally inspired by <a href="https://github.com/rll/rllab">rllab</a> (for example the logger remains nearly a direct copy).</p>
<p>Other published research code bases include <a href="https://github.com/openai/baselines">OpenAI Baselines</a> and <a href="https://github.com/google/dopamine">Dopamine</a>, both of which are implemented in Tensorflow, and neither of which are optimized to the extent of rlpyt nor contain all three algorithm families. <a href="https://ray.readthedocs.io/en/latest/rllib.html">Rllib</a>, built on top of <a href="https://github.com/ray-project/ray">Ray</a>, takes a different approach focused on distributed computing, possibly complicating small experiments. <a href="https://github.com/facebookresearch/Horizon">Facebook Horizon</a> offers a subset of algorithms and focuses on applications toward production at scale. In sum, rlpyt provides modular implementations of more algorithms and modular infrastructure for parallelism, making it a distinct toolset supporting a wide range of research uses.</p>
<p>We hope that rlpyt can facilitate easier use of existing deep RL techniques and serve as a launching point for research into new ones. For example, the more advanced topics of meta-learning, model-based, and multi-agent RL are not explicitly addressed in rlpyt, but applicable code components may still be helpful in accelerating their development. We expect the offerings of algorithms to grow over time. Let us know in the comments below if you have any questions or suggestions, check out the code with examples at <a href="https://github.com/astooke/rlpyt">https://github.com/astooke/rlpyt</a>, and read the <a href="https://arxiv.org/abs/1909.01500">white paper on arxiv</a> for more implementation and usage details. Happy reinforcement learning!</p>
</article>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via The Berkeley Artificial Intelligence Research Blog https://ift.tt/2EPVVfk<br>
<br>
September 27, 2019 at 11:55AM
